### ISSUE TITLE
Database migration strategy

### ISSUE THREAD

**Issue by kfangw** [bosnet/sebak#842]
- Related Issue
#834 

- Background
* Invariant `internal` data structure is only Block and Transaction
* Block
    * 	Version
    *	PrevBlockHash
    *	TransactionsRoot
    *	ProposedTime
    *	Height
    *	TotalTxs
    *	TotalOps
* Transaction
    *  Source
    *  Fee
    *  SequenceID
    *  Operations

Except above data structure and its fields, every other can be changed.
IMHO,
BlockAccount and BlockOperation data structure are the caches for speed up getting data.
It means that these data structure can be changed anytime.
For example, Index or model for the data.

To support this, we need a database migration strategy.
Here is a suggestion. It is originated from @anarcher 's idea

* The database structure is changed
* Launch a node as `watcher` mode with the change.
* Every validator is watched by at least one watcher node
* Wait until the watcher node sync all data from validator node in the changed data structure
* Kill the validator and elaborate the watcher node to the validator

I want to listen to your opinion that this scenario can be possible or not.
@spikeekips @anarcher @zzim2x @outsideris @Geod24 

Especially @zzim2x 's opinion is very crucial.

**Comment by spikeekips** [comment]
At first, 👍 .

This suggestion is focused on the maintenance by us, if watcher node operator see the db migration, the story should be changed. Of course, they can use this way, but we should provide the explicit way for them, I think.

This suggested way and plus, I suggest new sub command, `$ sebak db migrate` command, it checks the existing db status and if needed, migrate to the latest version.

**Comment by kfangw** [comment]
> At first, 👍 .
> 
> This suggestion is focused on the maintenance by us, if watcher node operator see the db migration, the story should be changed. Of course, they can use this way, but we should provide the explicit way for them, I think.
> 
> This suggested way and plus, I suggest new sub command, `$ sebak db migrate` command, it checks the existing db status and if needed, migrate to the latest version.

`sebak db migrate` command is good 👍 

To do, we need to manage and refactor the `block` package to be more suitable for this purpose.

**Comment by Geod24** [comment]
> Except above data structure and its fields, every other can be changed.

Not exactly. As I said in #834 (this issue could have been a comment in #834 BTW), change to a field type might be problematic as well. Take https://github.com/bosnet/sebak/pull/836 for example. It changes a type from string to an enum, but provide conversion routines. At the moment I do not know how LevelDB will react to this. If LevelDB has a proper support for the standard interfaces, and a relaxed policy, it could just work. But we have no guarantee of it.

What I think we should do is:
- Develop our API server, to reduce the amount of changes needed. This will make changes in the database in Sebak much more rare, as most changes we've been doing have because of the API. It will also allow for faster development, because we'll know we don't have to care about API server DB migration.
- Provide serialization / deserialization interfaces for Sebak. Something that guarantees we can convert from an old record to a new record.
- **Optionally**: Provide a global database version, for expensive operation. E.g. if we add a new field based on some block data, we'll have to reconstruct the table for the whole blockchain (with a stable blockchain, it is unlikely, but still possible).


I put those thing by order of priority. IMO the first item is the most important one. It will make most of the changes harmless (not breaking the database).

**Comment by spikeekips** [comment]
@Geod24 At this time, we can put the version in storage.
